### PR TITLE
fix: page.goto should resolve to response for self request (#1391)

### LINKED ISSUE #1391: Page.goto returns null for some urls
This issue extracts relevant informations from #1056 which was not very clear.

When used will request inception, puppeteer returns a `null` value sometimes. But the [documentation](https://github.com/GoogleChrome/puppeteer/blob/99103cbb97e6fe80de86c3001195355c3d8f51e0/docs/api.md#pagegotourl-options) says:

> **NOTE** `page.goto` either throw or return a main resource response. The only exception is navigation to `about:blank`, which would succeed and return `null`.

### Steps to reproduce

**Tell us about your environment:**

* Puppeteer version: 99103cbb97e6fe80de86c3001195355c3d8f51e0
* Platform / OS version: debian stretch
* URLs (if applicable): 
  - http://giffysk8s.blogspot.com/
  - http://lemaninignesinden.blogspot.com/2015/02/vitaminenzimklorofilsifa.html
  - http://henke-s.blogspot.com/
  - http://www.cometogetherkids.com/2011/03/fleece-flower-petal-pillows.html
  - http://www.wildstar-online.com/uk/drops/1/strain/
  - https://www.vanityfair.com/culture/photos/2010/06/world-cup-portfolio-201006
  - http://www.sportsnet.ca/baseball/mlb/blue-jays-exploring-trade-options-before-free-agency/
  - http://www.fc-weisweil.de/

**What steps will reproduce the problem?**

1. save the code bellow as `test.js`
2. run `nodejs test.js "http://giffysk8s.blogspot.com/"`
3. check for a line with `result: null` in the output. It is probably the last line.
4. comment the line 40 `blockImages(page);`
5. run `nodejs test.js "http://giffysk8s.blogspot.com/"`
6. enjoy the correct result object

`test.js`:

```javascript
'use strict';

const puppeteer = require('puppeteer');

const URL = process.argv[2];

async function blockImages(page) {
  await page.setRequestInterception(true);
  page.on("request", (request) => {
    if (request.resourceType === "image") {
      request.abort();
    } else {
      request.continue();
    }
  });
}

(async () => {
  console.log("processing url", URL);
  process.on("uncaughtException", (e) => {
    console.error("Unhandled exeption:", e);
    process.exit(1);
  });
  process.on("unhandledRejection", (reason, p) => {
    console.error("Unhandled Rejection at: Promise", p, "reason:", reason);
    process.exit(2);
  });
  const args = [
    "--disable-setuid-sandbox",
    "--no-sandbox",
  ];
  const options = {
    args,
    headless: true,
    ignoreHTTPSErrors: true,
    dumpio: true,
  };
  const browser = await puppeteer.launch(options);
  const page = await browser.newPage();
  blockImages(page);
  const res = await page.goto(URL, { timeout: 30000, waitUntil: "load" });
  console.log("result:", res);
  await page.close();
  await browser.close();
})();
```

**What is the expected result?**

A normal result object should be printed.

```
$ nodejs test.js "http://giffysk8s.blogspot.com/"
processing url http://giffysk8s.blogspot.com/
[1115/065856.610648:ERROR:gpu_process_transport_factory.cc(1009)] Lost UI shared context.
[1115/065856.637350:ERROR:instance.cc(49)] Unable to locate service manifest for metrics
[1115/065856.637370:ERROR:service_manager.cc(889)] Failed to resolve service name: metrics

DevTools listening on ws://127.0.0.1:42869/devtools/browser/753f024d-a38a-409e-95aa-64165a012e80
[1115/065856.752637:ERROR:nss_util.cc(724)] After loading Root Certs, loaded==false: NSS error code: -8018
[1115/065858.367284:INFO:CONSOLE(26)] "Mixed Content: The page at 'https://www.blogger.com/navbar.g?targetBlogID=4652634595166455172&blogName=Canvasses+of+Poetry+and+Prose&publishMode=PUBLISH_MODE_BLOGSPOT&navbarType=BLUE&layoutType=LAYOUTS&searchRoot=http://giffysk8s.blogspot.com/search&blogLocale=en&v=2&homepageUrl=http://giffysk8s.blogspot.com/&vt=-6074931036815180963&usegapi=1&jsh=m%3B%2F_%2Fscs%2Fapps-static%2F_%2Fjs%2Fk%3Doz.gapi.en_GB.eAe10hJSHzc.O%2Fm%3D__features__%2Fam%3DAQ%2Frt%3Dj%2Fd%3D1%2Frs%3DAGLTcCOSrQAixLCyS0W7RP8OLBQKClcz2w#id=navbar-iframe&_gfid=navbar-iframe&parent=http%3A%2F%2Fgiffysk8s.blogspot.sg&pfname=&rpctoken=34501815' was loaded over a secure connection, but contains a form that targets an insecure endpoint 'http://giffysk8s.blogspot.com/search'. This endpoint should be made available over a secure connection.", source: https://www.blogger.com/navbar.g?targetBlogID=4652634595166455172&blogName=Canvasses+of+Poetry+and+Prose&publishMode=PUBLISH_MODE_BLOGSPOT&navbarType=BLUE&layoutType=LAYOUTS&searchRoot=http://giffysk8s.blogspot.com/search&blogLocale=en&v=2&homepageUrl=http://giffysk8s.blogspot.com/&vt=-6074931036815180963&usegapi=1&jsh=m%3B%2F_%2Fscs%2Fapps-static%2F_%2Fjs%2Fk%3Doz.gapi.en_GB.eAe10hJSHzc.O%2Fm%3D__features__%2Fam%3DAQ%2Frt%3Dj%2Fd%3D1%2Frs%3DAGLTcCOSrQAixLCyS0W7RP8OLBQKClcz2w#id=navbar-iframe&_gfid=navbar-iframe&parent=http%3A%2F%2Fgiffysk8s.blogspot.sg&pfname=&rpctoken=34501815 (26)
[1115/065858.447190:WARNING:render_frame_host_impl.cc(2679)] OnDidStopLoading was called twice.
[1115/065858.447639:WARNING:render_frame_host_impl.cc(2679)] OnDidStopLoading was called twice.
[1115/065858.871226:WARNING:render_frame_host_impl.cc(2679)] OnDidStopLoading was called twice.
result: Response {
  _client: 
   Session {
     domain: null,
     _events: 
      { 'Page.frameAttached': [Function],
        'Page.frameNavigated': [Function],
        'Page.frameDetached': [Function],
        'Runtime.executionContextCreated': [Function],
        'Page.lifecycleEvent': [Function],
        'Network.requestWillBeSent': [Function: bound _onRequestWillBeSent],
        'Network.requestIntercepted': [Function: bound _onRequestIntercepted],
        'Network.responseReceived': [Function: bound _onResponseReceived],
        'Network.loadingFinished': [Function: bound _onLoadingFinished],
        'Network.loadingFailed': [Function: bound _onLoadingFailed],
        'Page.loadEventFired': [Function],
        'Runtime.consoleAPICalled': [Function],
        'Page.javascriptDialogOpening': [Function],
        'Runtime.exceptionThrown': [Function],
        'Security.certificateError': [Function],
        'Inspector.targetCrashed': [Function],
        'Performance.metrics': [Function] },
     _eventsCount: 17,
     _maxListeners: undefined,
     _lastId: 11,
     _callbacks: Map {},
     _connection: 
      Connection {
        domain: null,
        _events: [Object],
        _eventsCount: 3,
        _maxListeners: undefined,
        _url: 'ws://127.0.0.1:42869/devtools/browser/753f024d-a38a-409e-95aa-64165a012e80',
        _lastId: 14,
        _callbacks: Map {},
        _delay: 0,
        _ws: [Object],
        _sessions: [Object],
        _closeCallback: [Function] },
     _targetId: '(63A63DF21210058BAD5DE73B52FAD0A8)',
     _sessionId: '(63A63DF21210058BAD5DE73B52FAD0A8):1' },
  _request: 
   Request {
     _client: 
      Session {
        domain: null,
        _events: [Object],
        _eventsCount: 17,
        _maxListeners: undefined,
        _lastId: 11,
        _callbacks: Map {},
        _connection: [Object],
        _targetId: '(63A63DF21210058BAD5DE73B52FAD0A8)',
        _sessionId: '(63A63DF21210058BAD5DE73B52FAD0A8):1' },
     _requestId: '11529.1',
     _interceptionId: null,
     _allowInterception: false,
     _interceptionHandled: false,
     _response: [Circular],
     _failureText: null,
     _completePromiseFulfill: [Function],
     _completePromise: Promise { undefined },
     url: 'http://giffysk8s.blogspot.sg/',
     resourceType: 'document',
     method: 'GET',
     postData: undefined,
     headers: 
      { 'user-agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) HeadlessChrome/64.0.3264.0 Safari/537.36' } },
  _contentPromise: null,
  status: 200,
  ok: true,
  url: 'http://giffysk8s.blogspot.sg/',
  headers: 
   { date: 'Wed, 15 Nov 2017 06:58:57 GMT',
     'content-encoding': 'gzip',
     'x-content-type-options': 'nosniff',
     'last-modified': 'Thu, 08 Sep 2016 04:31:00 GMT',
     server: 'GSE',
     etag: 'W/"30bf6632614ccc69849f4627193b6855b9928f434964bcc926be27de576b7652"',
     'content-type': 'text/html; charset=UTF-8',
     'cache-control': 'private, max-age=0',
     'content-length': '19943',
     'x-xss-protection': '1; mode=block',
     expires: 'Wed, 15 Nov 2017 06:58:57 GMT' } }
```

**What happens instead?**

```
$ nodejs test.js "http://giffysk8s.blogspot.com/"
processing url http://giffysk8s.blogspot.com/
[1115/065217.501797:ERROR:gpu_process_transport_factory.cc(1009)] Lost UI shared context.
[1115/065217.532286:ERROR:instance.cc(49)] Unable to locate service manifest for metrics
[1115/065217.532305:ERROR:service_manager.cc(889)] Failed to resolve service name: metrics

DevTools listening on ws://127.0.0.1:40731/devtools/browser/c7ec5251-d794-49a1-8fd9-f26c94aa9c39
[1115/065217.652862:ERROR:nss_util.cc(724)] After loading Root Certs, loaded==false: NSS error code: -8018
[1115/065220.088205:WARNING:render_frame_host_impl.cc(2679)] OnDidStopLoading was called twice.
[1115/065220.088888:WARNING:render_frame_host_impl.cc(2679)] OnDidStopLoading was called twice.
[1115/065220.275850:INFO:CONSOLE(26)] "Mixed Content: The page at 'https://www.blogger.com/navbar.g?targetBlogID=4652634595166455172&blogName=Canvasses+of+Poetry+and+Prose&publishMode=PUBLISH_MODE_BLOGSPOT&navbarType=BLUE&layoutType=LAYOUTS&searchRoot=http://giffysk8s.blogspot.com/search&blogLocale=en&v=2&homepageUrl=http://giffysk8s.blogspot.com/&vt=-6074931036815180963&usegapi=1&jsh=m%3B%2F_%2Fscs%2Fapps-static%2F_%2Fjs%2Fk%3Doz.gapi.en_GB.eAe10hJSHzc.O%2Fm%3D__features__%2Fam%3DAQ%2Frt%3Dj%2Fd%3D1%2Frs%3DAGLTcCOSrQAixLCyS0W7RP8OLBQKClcz2w#id=navbar-iframe&_gfid=navbar-iframe&parent=http%3A%2F%2Fgiffysk8s.blogspot.sg&pfname=&rpctoken=18836155' was loaded over a secure connection, but contains a form that targets an insecure endpoint 'http://giffysk8s.blogspot.com/search'. This endpoint should be made available over a secure connection.", source: https://www.blogger.com/navbar.g?targetBlogID=4652634595166455172&blogName=Canvasses+of+Poetry+and+Prose&publishMode=PUBLISH_MODE_BLOGSPOT&navbarType=BLUE&layoutType=LAYOUTS&searchRoot=http://giffysk8s.blogspot.com/search&blogLocale=en&v=2&homepageUrl=http://giffysk8s.blogspot.com/&vt=-6074931036815180963&usegapi=1&jsh=m%3B%2F_%2Fscs%2Fapps-static%2F_%2Fjs%2Fk%3Doz.gapi.en_GB.eAe10hJSHzc.O%2Fm%3D__features__%2Fam%3DAQ%2Frt%3Dj%2Fd%3D1%2Frs%3DAGLTcCOSrQAixLCyS0W7RP8OLBQKClcz2w#id=navbar-iframe&_gfid=navbar-iframe&parent=http%3A%2F%2Fgiffysk8s.blogspot.sg&pfname=&rpctoken=18836155 (26)
[1115/065220.546524:WARNING:render_frame_host_impl.cc(2679)] OnDidStopLoading was called twice.
result: null
```

### PR DESCRIPTION
**Pull request by yujiosaka**
I confirmed that the following URLs will start resolving responses by this fix:

- http://giffysk8s.blogspot.com/
- http://lemaninignesinden.blogspot.com/2015/02/vitaminenzimklorofilsifa.html
- http://henke-s.blogspot.com/
- http://www.qh119.com/10022/
- http://corralitopiraeus.blogspot.com/2017/07/malma-con-piezas-de-ordenador-y.html
- http://kedm.org/

I cannot repro for the following URLs:

- http://www.cometogetherkids.com/2011/03/fleece-flower-petal-pillows.html
- http://www.wildstar-online.com/uk/drops/1/strain/
- https://www.vanityfair.com/culture/photos/2010/06/world-cup-portfolio-201006
- http://www.sportsnet.ca/baseball/mlb/blue-jays-exploring-trade-options-before-free-agency/
- http://www.fc-weisweil.de/
- https://www.lovedignity.com/ex-back-experts-review/
- http://www.hainanrra.com/bjbw/
- http://www.manlexi.com/rqfp/
- http://www.12011.org/ljnn/
- http://www.grgds.com/jbs/

But these may have been fixed already like https://github.com/GoogleChrome/puppeteer/issues/631 .

Fixes: https://github.com/GoogleChrome/puppeteer/issues/1391

@aslushnikov

### CODE REVIEW

**Comment by aslushnikov** [comment]
@yujiosaka thank you for the patch! The only thing I don't understand is how come this is related to request interception. Is this just a timing issue?

**Comment by yujiosaka** [comment]
@aslushnikov 
It's a very good point. I assume it was just a timing issue, and the same problem may occur without request interception.
While investigating this problem, I noticed that the request to the same url occurs after load event. It may be just that `requestIntercepted` is fired before `requestWillBeSent`.

If that's the case, I can add another test without request interception.
Let me see.

**Comment by yujiosaka** [comment]
> Is this just a timing issue?

It was right! I could repro the same problem even without request interception with [this small html](https://github.com/GoogleChrome/puppeteer/pull/1781/files#diff-a8640a3e711f8016f0576102f097cfc4R1).

Thus, I added a test for non request interception version. https://github.com/GoogleChrome/puppeteer/pull/1781/commits/37f0d67d68798afc608ba0bce44916862bb08ff4

Thank you for pointing it out!

File: test/test.js [hunk 1]
```diff
@@ -1397,6 +1402,13 @@ describe('Page', function() {
       expect(requests.length).toBe(5);
       expect(requests[2].resourceType()).toBe('document');
     });
+    it('should work with self requesting page', async({page, server}) => {
```
Comment (aslushnikov) [inline test/test.js]: Can you please drop this test? It doesn't cover anything new wrt to another one you've added here.

Comment (yujiosaka) [reply test/test.js]: Do you think so? I would like to hear more reasons.

With `page.setRequestInterception(true)` it goes different code path.
I thought that's the reason why we have such tests as redirect chain for both non interception and interception versions.

Comment (aslushnikov) [reply test/test.js]: > With page.setRequestInterception(true) it goes different code path.

@yujiosaka from the point of view of Page class and `page.goto` method specifically, there's no difference in the codepath. Or am I missing something?

Comment (yujiosaka) [reply test/test.js]: > Or am I missing something?

No! I understand your point, so I'll drop the test :)

Comment (yujiosaka) [reply test/test.js]: Done!
https://github.com/GoogleChrome/puppeteer/pull/1781/commits/83daf475a75dbe4344996a6a87be2e52269bccce

**Review by aslushnikov — APPROVED** [review]

**Pull request merged**